Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -298,157 +321,309 @@ void GCMemcardManager::SetSlotFileInteractive(int slot)
+          GetFormatDescription(Memcard::SavefileFormat::GCS) + QStringLiteral(" (*.gcs);;") +
+          GetFormatDescription(Memcard::SavefileFormat::SAV) + QStringLiteral(" (*.sav);;") +
+          tr("All Files") + QStringLiteral(" (*)"));
 
-  // Ask for confirmation if we are to delete multiple files
-  if (count > 1)
-  {
-    QString text = count == 1 ? tr("Do you want to delete the selected save file?") :
-                                tr("Do you want to delete the %1 selected save files?").arg(count);
+  if (paths.isEmpty())
+    return;
 
-    auto response = ModalMessageBox::question(this, tr("Question"), text);
-    ;
+  std::vector<Memcard::Savefile> savefiles;
+  savefiles.reserve(paths.size());
 
-    if (response == QMessageBox::Abort)
+  for (const QString& path : paths)
+  {
+    auto read_result = Memcard::ReadSavefile(path.toStdString());
+    auto* savefile = std::get_if<Memcard::Savefile>(&read_result);
+    if (savefile)
+    {
+      savefiles.emplace_back(std::move(*savefile));
+      continue;
+    }
+
+    const auto* error_code = std::get_if<Memcard::ReadSavefileErrorCode>(&read_result);
+    if (error_code)
+    {
+      ModalMessageBox::warning(this, tr("Import Failed"),
+                               tr("Failed to import \"%1\".").arg(path) + QStringLiteral("\n") +
+                                   GetErrorMessageForErrorCode(*error_code));
       return;
+    }
+
+    // should never get here but just in case...
+    ModalMessageBox::warning(this, tr("Import Failed"), tr("Failed to import \"%1\".").arg(path));
+    return;
   }
 
-  std::vector<int> file_indices;
-  for (int i = 0; i < count; i++)
+  ImportFiles(m_active_slot, savefiles);
+}
+
+void GCMemcardManager::CopyFiles()
+{
+  const auto& source_card = m_slot_memcard[m_active_slot];
+  if (!source_card)
+    return;
+
+  auto& target_card = m_slot_memcard[!m_active_slot];
+  if (!target_card)
+    return;
+
+  const auto selected_indices = GetSelectedFileIndices();
+  if (selected_indices.empty())
+    return;
+
+  const auto savefiles = Memcard::GetSavefiles(*source_card, selected_indices);
+  if (savefiles.empty())
   {
-    auto sel = selection[i * m_slot_table[m_active_slot]->columnCount()];
-    file_indices.push_back(memcard->GetFileIndex(m_slot_table[m_active_slot]->row(sel)));
+    ModalMessageBox::warning(this, tr("Copy Failed"),
+                             tr("Failed to read selected savefile(s) from memory card."));
+    return;
   }
 
-  for (int file_index : file_indices)
+  ImportFiles(!m_active_slot, savefiles);
+}
+
+void GCMemcardManager::DeleteFiles()
+{
+  auto& card = m_slot_memcard[m_active_slot];
+  if (!card)
+    return;
+
+  const auto selected_indices = GetSelectedFileIndices();
+  if (selected_indices.empty())
+    return;
+
+  const QString text =
+      selected_indices.size() == 1 ?
+          tr("Do you want to delete the selected save file?") :
```

Comment:
The way to handle plurals here is to do it in a single statement like:

```cpp
tr("Do you want to delete the %n selected save file(s)?", "", static_cast<int>(selected_indices.size()));
```

This allows languages with differing handling of quantities to be translated better. (see the handling plurals part of Qt localizing [docs](https://doc.qt.io/qt-5/i18n-source-translation.html#handling-plurals))

Reply:
Ah, I didn't know you could do that! That probably applies in a few more places here too.